csrc/ir/base_nodes.cpp

```diff
@@ -116,32 +116,60 @@ bool Val::removeUse(Expr* expr) {
   return false;
 }
 
+bool Val::sameVal(const Val* other_val) const {
+  if (other_val == nullptr) {
+    return false;
+  }
+  if (typeid(*this) != typeid(*other_val)) {
+    return false;
+  }
+  if ((definition_ == nullptr) != (other_val->definition_ == nullptr)) {
+    return false;
+  }
+  if (vtype_ != other_val->vtype_) {
+    return false;
+  }
+  if (dtype_ != other_val->dtype_) {
+    return false;
+  }
+  if (value_.hasValue() != other_val->value_.hasValue()) {
+    return false;
+  }
+  if (definition_ == nullptr) {
+    // Check this condition in Val::sameAs and Val::sameDefinition
+    return true;
```

Comment:
nitpick: returning `true` for sameAs when this->definition_ == nullptr felt very strange and wrong. I think we should refactor this utility to `notSameVal` instead.

Reply:
This check means it might be the same val and requires further checking in `Val::sameAs` and `Val::sameDefinition`.

You're suggestion doesn't seem correct. If you return `false` in `notSameVal`, then, how will you know to check further in `Val::sameAs` or `Val::sameDefinition`.

Reply:
I'll rename it to `maybeSameVal`. It is a private helper function, so it cannot be mistakenly used.